packages/squiggle-lang/src/public/SqProject/SqModule.ts

```diff
@@ -0,0 +1,180 @@
+import { parse } from "../../ast/parse.js";
+import { AST, LocationRange } from "../../ast/types.js";
+import { Env } from "../../dists/env.js";
+import { errMap, result } from "../../utility/result.js";
+import {
+  SqCompileError,
+  SqError,
+  SqImportError,
+  SqOtherError,
+} from "../SqError.js";
+import { SqLinker } from "../SqLinker.js";
+import { ModulePointer, ProjectState } from "./ProjectState.js";
+import { SqModuleOutput } from "./SqModuleOutput.js";
+import { getHash } from "./utils.js";
+
+export type Import = {
+  name: string;
+  hash: string | undefined;
+  variable: string;
+  location: LocationRange;
+};
+
+export function importToPointer(importBinding: Import): ModulePointer {
+  return {
+    name: importBinding.name,
+    hash: importBinding.hash,
+  };
+}
+
+type ImportModules =
+  | {
+      type: "loaded";
+      value: Record<string, SqModule>;
+    }
+  | {
+      type: "loading";
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+type ImportOutputs =
+  | {
+      type: "loaded";
+      value: Record<string, SqModuleOutput>;
+    }
+  | {
+      type: "loading"; // or running - i.e. ImportModules might be loaded, but outputs don't exist yet
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+export class SqModule {
+  name: string;
```

Comment:
This is `sourceId`. Maybe I should've kept the old name, but:
- I did a few renames and rewrites, e.g. at some point this class was called `SqSource`, and it was awkward to have `source.sourceId`
- `module.name` is in context of `module` so it's ok that it's short; and if you extract it to a variable, you can name it however you like

In general, I avoid the word "source" in this PR; "source" sounds like "code string", and sometimes it's that, but in other times it's the module class instance, so it was confusing.

There are still some places where `sourceId` is used, we probably should clean them up later and rename to `moduleName`.

Reply:
I'm used to `name` existing mainly for users, not to act as a lookup in code. I'd expect it to be called `id` if the main use is for other modules to access it.

Reply:
I'd argue that `id` can lead to more confusion because this field is very much not a unique id in the context of `SqProject`.

In the normal editing workflow you'll often get multiple `SqModule` instances with different code but the same name; `name` will be the same but `hash` will be different.

So `name` really is, mostly, for users. The only exception is that sources refer to other sources by name, everything else uses hash-based lookups.